python/ray/data/lazy_dataset.py

```diff
@@ -0,0 +1,493 @@
+
+        def impl(self, *args, **kwargs) -> "LazyDatasetPipeline":
+            return LazyDatasetPipeline(
+                func=delegate, args=(self,) + args, kwargs=kwargs
+            )
+
+        impl.__name__ = delegate.__name__
+        lazy_preamble = """
+        Lazy version of ``DatasetPipeline.{method}``, returning a LazyDatasetPipeline.
+        """.format(
+            method=method
+        )
+        impl.__doc__ = lazy_preamble + delegate.__doc__
+        setattr(
+            impl,
+            "__signature__",
+            inspect.signature(delegate).replace(
+                return_annotation="LazyDatasetPipeline"
+            ),
+        )
+        return impl
+
+    setattr(LazyDatasetPipeline, method, make_impl(method))
+
+for method in PIPELINE_CONSUME_OPS:
+
+    def make_impl(method):
+        delegate = getattr(DatasetPipeline, method)
+
+        def impl(self, *args, **kwargs):
+            return getattr(self.compute(), method)(*args, **kwargs)
+
+        impl.__name__ = delegate.__name__
+        lazy_preamble = """
+        NOTE: This will trigger computation and materialize the lazy DatasetPipeline!
+        """
+        impl.__doc__ = lazy_preamble + delegate.__doc__
+        setattr(
+            impl,
+            "__signature__",
+            inspect.signature(delegate),
+        )
+        return impl
+
+    setattr(LazyDatasetPipeline, method, make_impl(method))
+
+
+def add_lazy_option(fn):
+    """
+    Add a lazy kwarg to a Dataset-generating function, instead returning a
+    LazyDataset.
+    """
+
+    @functools.wraps(fn)
+    def wrapper(*args, **kwargs):
+        if kwargs.pop("lazy", False):
+            return LazyDataset(func=fn, args=args, kwargs=kwargs)
+        return fn(*args, **kwargs)
+
+    doc = fn.__doc__
+    wrapper.__doc__ = _add_arg_to_docstring(doc)
+
+    return wrapper
+
+
+def _add_arg_to_docstring(doc):
+    lines = doc.split("\n")
+    i = 0
+    while i < len(lines):
+        if lines[i].strip().startswith("Args:"):
+            # Found the start of the args section.
+            break
+        i += 1
+    else:
+        # No args section found.
+        return doc
+    i += 1
+    # Get indentation.
+    indent = re.match(r"\s*", lines[i]).group(0)
+    while i < len(lines):
+        if len(lines[i]) == 0:
+            # Found the end of the args section (double-newline).
+            break
+        i += 1
+    else:
+        # No args section found.
+        return doc
+    lazy_arg_doc = (
+        "lazy: Whether to create the dataset lazily. If True, a LazyDataset will be "
+        "returned and the dataset won't be created and processed until .compute() is "
+        "called or it is consumed. See the LazyDataset docs for more information."
+    )
+    wrapper = textwrap.TextWrapper(
+        width=80, initial_indent=indent, subsequent_indent=indent + "    "
+    )
+    lines.insert(i, wrapper.fill(lazy_arg_doc))
+    return "\n".join(lines)
```

Comment:
Haven't tested this arg docstring injection, but the general approach seems doable.